src/rmkit/ui/events.cpy

```diff
@@ -0,0 +1,184 @@
+        switch state:
+          case IDLE:              return IDLE_down(ev)
+          case WAIT_DOUBLE_CLICK: return WAIT_DOUBLE_CLICK_down(ev)
+          default:                return reset()
+      }
+      mouse->up += PLS_LAMBDA(auto &ev) {
+        switch state:
+          case DOWN:        return DOWN_up(ev)
+          case DRAGGING:    return finish(drag_end, ev)
+          case SECOND_DOWN: return finish(double_click, ev)
+          default:          return reset()
+      }
+      mouse->move += PLS_LAMBDA(auto &ev) {
+        switch state:
+          case DOWN:        return DOWN_move(ev)
+          case DRAGGING:    return DRAGGING_move(ev)
+          default:          return
+      }
+      mouse->leave += PLS_LAMBDA(auto &ev) {
+        switch state:
+          case DRAGGING:    return finish(drag_end, ev)
+          default:          return reset()
+      }
+
+    protected:
+    input::SynMotionEvent prev_ev
+    TimerPtr long_press_timer
+    TimerPtr single_click_timer
+    enum STATE { IDLE, DOWN, DRAGGING, WAIT_DOUBLE_CLICK, SECOND_DOWN }
+    STATE state = IDLE
+
+    inline void IDLE_down(input::SynMotionEvent &ev):
+      reset()
+      state = DOWN
+      prev_ev = ev
+      if not long_press.empty():
+        long_press_timer = ui::set_timeout([=]() {
+          auto ev_copy = ev
+          finish(long_press, ev_copy)
+        }, long_press_timeout)
+
+    inline void DOWN_up(input::SynMotionEvent &ev):
+      cancel_long_press()
+      if double_click.empty():
+        finish(single_click, ev)
+      else:
+        state = WAIT_DOUBLE_CLICK
+        single_click_timer = ui::set_timeout([=]() {
+          auto ev_copy = ev
+          finish(single_click, ev_copy)
+        }, double_click_timeout)
+
+    inline void DOWN_move(input::SynMotionEvent &ev):
+      if outside_tolerance(ev, touch_slop_size):
+        state = DRAGGING
+        cancel_long_press()
+        drag_start(prev_ev)
+        prev_ev = ev
+
+    inline void WAIT_DOUBLE_CLICK_down(input::SynMotionEvent &ev):
+      state = SECOND_DOWN
+      cancel_single_click()
+      prev_ev = ev
+
+    inline void DRAGGING_move(input::SynMotionEvent &ev):
+      if outside_tolerance(ev, dragging_step_size):
+        dragging(ev)
+        prev_ev = ev
+
+    // helpers
+    void reset():
+      cancel_long_press()
+      cancel_single_click()
+      state = IDLE
+
+    inline void finish(MOUSE_EVENT & handler, input::SynMotionEvent & ev):
+      handler(ev)
+      reset()
+
+    void cancel_long_press()
+      if long_press_timer:
+        ui::cancel_timer(long_press_timer)
+        long_press_timer = nullptr
+
+    void cancel_single_click()
+      if single_click_timer:
+        ui::cancel_timer(single_click_timer)
+        single_click_timer = nullptr
+
+    bool outside_tolerance(const input::SynMotionEvent & ev, int tolerance):
+     return std::abs(ev.x - prev_ev.x) > tolerance || std::abs(ev.y - prev_ev.y) > tolerance
```

Comment:
good catch, thanks -- like I said, still figuring out the cpy auto-indentation 🙂 (your cpy syntax file from earlier helped, but I'm still not quite there)

Reply:
for indentation, when i use `o`, or `O`, vim opens up a line with the same indent as the current one. i have to press tab to add in two spaces, so i guess i don't have auto-indentation working (it would automatically figure out the number of spaces based on if the current line ends with a ':'?)

Reply:
ah ok, I'm just using ft=cpp, so I also needed to turn off most of the smart indentation options. I think I figured out how to get to where you are w/ indentation (I have to manually indent, but new lines start at the same level), which is much better!